### ISSUE TITLE
example10/conduit.c: make error in functions flux_respond, flux_kvs_commit 

### ISSUE THREAD

**Issue by cmoussa1** [flux-framework/flux-workflow-examples#31]
When running `make` on **conduit.c** with the new execution system, I get the following error stack trace:

```
bash-4.2$ export PKG_CONFIG_PATH='/usr/global/tools/flux/toss_3_x86_64_ib/new/lib/pkgconfig' 
bash-4.2$ make
gcc -I/collab/usr/global/tools/flux/toss_3_x86_64_ib/flux-0.13.x-20191111/include   conduit.c -fPIC -c -o conduit.o
conduit.c: In function 'conduit_put_request_cb':
conduit.c:90:31: warning: passing argument 3 of 'flux_respond' makes pointer from integer without a cast
     if (flux_respond (h, msg, errno, NULL) < 0)
                               ^
In file included from /collab/usr/global/tools/flux/toss_3_x86_64_ib/flux-0.13.x-20191111/include/flux/core/flux.h:21:0,
                 from /collab/usr/global/tools/flux/toss_3_x86_64_ib/flux-0.13.x-20191111/include/flux/core.h:18,
                 from conduit.c:9:
/collab/usr/global/tools/flux/toss_3_x86_64_ib/flux-0.13.x-20191111/include/flux/core/response.h:68:5: note: expected 'const char *' but argument is of type 'int'
 int flux_respond (flux_t *h, const flux_msg_t *request, const char *s);
     ^
conduit.c:90:9: error: too many arguments to function 'flux_respond'
     if (flux_respond (h, msg, errno, NULL) < 0)
         ^
In file included from /collab/usr/global/tools/flux/toss_3_x86_64_ib/flux-0.13.x-20191111/include/flux/core/flux.h:21:0,
                 from /collab/usr/global/tools/flux/toss_3_x86_64_ib/flux-0.13.x-20191111/include/flux/core.h:18,
                 from conduit.c:9:
/collab/usr/global/tools/flux/toss_3_x86_64_ib/flux-0.13.x-20191111/include/flux/core/response.h:68:5: note: declared here
 int flux_respond (flux_t *h, const flux_msg_t *request, const char *s);
     ^
conduit.c: In function 'mod_main':
conduit.c:162:28: warning: passing argument 3 of 'flux_kvs_commit' makes integer from pointer without a cast
     flux_kvs_commit (h, 0, txn);
                            ^
In file included from /collab/usr/global/tools/flux/toss_3_x86_64_ib/flux-0.13.x-20191111/include/flux/core/kvs.h:20:0,
                 from /collab/usr/global/tools/flux/toss_3_x86_64_ib/flux-0.13.x-20191111/include/flux/core.h:19,
                 from conduit.c:9:
/collab/usr/global/tools/flux/toss_3_x86_64_ib/flux-0.13.x-20191111/include/flux/core/kvs_commit.h:22:16: note: expected 'int' but argument is of type 'struct flux_kvs_txn_t *'
 flux_future_t *flux_kvs_commit (flux_t *h, const char *ns, int flags,
                ^
conduit.c:162:5: error: too few arguments to function 'flux_kvs_commit'
     flux_kvs_commit (h, 0, txn);
     ^
In file included from /collab/usr/global/tools/flux/toss_3_x86_64_ib/flux-0.13.x-20191111/include/flux/core/kvs.h:20:0,
                 from /collab/usr/global/tools/flux/toss_3_x86_64_ib/flux-0.13.x-20191111/include/flux/core.h:19,
                 from conduit.c:9:
/collab/usr/global/tools/flux/toss_3_x86_64_ib/flux-0.13.x-20191111/include/flux/core/kvs_commit.h:22:16: note: declared here
 flux_future_t *flux_kvs_commit (flux_t *h, const char *ns, int flags,
                ^
make: *** [conduit.o] Error 1

```

**Comment by dongahn** [comment]
@cmoussa1: Yeah I think this needs to be updated to the new API changes.  

https://github.com/flux-framework/flux-core/blob/master/doc/man3/flux_respond.adoc has the latest definitions of flux_respond and its variant for example.

**Comment by dongahn** [comment]
Something like the following should work. Code untested but you get the idea.

```c++
/* request callback called when conduit.put request is invoked */
static void conduit_put_request_cb (flux_t *h, flux_msg_handler_t *w,
                                    const flux_msg_t *msg, void *arg)
{
    int rc = -1;
    const char *topic = NULL;
    struct conduit_ctx *ctx = getctx (h);
    const char *data = NULL;

    flux_log (h, LOG_INFO, "conduit_put_request_cb:");
    if (ctx->connected == false) {
        flux_log (h, LOG_INFO, "conduit not connected");
        errno = ENOTCONN;
        goto error;
    }
    if (flux_request_unpack (msg, &topic, "{s:s}", "data", &data)) {
        flux_log_error (h, "%s", __FUNCTION__);
        goto error;
    }
    if (conduit_send (h, data) < 0) {
        errno = EPROTO;
        goto error;
    }
    if (flux_respond (h, msg, NULL) < 0)
        flux_log_error (h, "%s: flux_respond", __FUNCTION__);
    return;
error:
    if (flux_respond_error (h, msg, errno, NULL) < 0)
        flux_log_error (h, "%s: flux_respond_error", __FUNCTION__);
}
```

**Comment by dongahn** [comment]
The man page for `flux_kvs_commit` is https://github.com/flux-framework/flux-core/blob/master/doc/man3/flux_kvs_commit.adoc.

**Comment by dongahn** [comment]
Something like the following should work?

```c++
    /* Put the rank where this module is loaded into conduit key
     */
    flux_kvs_txn_t *txn = flux_kvs_txn_create ();
    flux_kvs_txn_pack (txn, 0, "conduit", "i", rank);

    flux_future_t *f = NULL;
    if (!(f = flux_kvs_commit (ctx->h, NULL, 0, txn))) {
        flux_log_error (...);
        goto done;
    }
    flux_future_wait_for (f, -1.0);
    flux_future_destroy (f);
    flux_kvs_txn_destroy (txn);
```

**Comment by cmoussa1** [comment]
Thanks for your help @dongahn. Your suggestions fixed the `make` problem; I was able to launch a Flux instance and submit `datastore.py`, but it results in a hang. Going to explore this now.

**Comment by dongahn** [comment]
Thank you for the update. I vaguely remember there was some problems with Flux's printing outputs. Back then, it wasn't an actual hang but a problem w/ output printing. This could be a completely different problem though.

**Comment by cmoussa1** [comment]
Gotcha. Just to add some more info, when `mini submit`ting **datastore.py**, it will run until the job exceeds its time limit and its allocation will be revoked:

```
bash-4.2$ flux job attach 219227881472
salloc: Job 3927768 has exceeded its time limit and its allocation has been revoked.
srun: Job step aborted: Waiting up to 62 seconds for job step to finish.
srun: Job step aborted: Waiting up to 62 seconds for job step to finish.
```

**Comment by dongahn** [comment]
What does `flux job list` say about the datastore.py job? Maybe it is not even scheduled?

**Comment by grondo** [comment]
Try running `flux job attach -XE <jobid>` to see the job and execution events printed in real time. That may help narrow down the issue. You may also want to add some debug output to `datastore.py` to see what it is doing.

**Comment by cmoussa1** [comment]
@dongahn Running `flux job list` on **datastore.py** shows that it runs:

```
bash-4.2$ flux job list                   
JOBID		STATE	USERID	PRI	T_SUBMIT
182770991104	R	58985	16	2019-11-15T17:12:42Z
```

@grondo when running `flux job attach -XE <jobid>`, **datastore.py** gets here:

```
0.000s: job.submit {"userid":58985,"priority":16,"flags":0}
0.016s: job.depend
0.024s: job.alloc {"note":""}
0.044s: job.start
0.026s: exec.init
0.031s: exec.starting
0.043s: exec.running
0.173s: exec.input-ready {"leader-rank":0}
0.173s: exec.output-ready
```

However, a pretty frequent problem I'm running into when trying to re-walk through this example is either a seg fault or a failed assertion:

```
bash-4.2$ flux mini submit -N 1 -n 1 ./datastore.py 
326400737280
bash-4.2$ srun: error: quartz19: task 0: Segmentation fault (core dumped)
srun: First task exited 30s ago
srun: step:3930466.1 tasks 1-2: running
srun: step:3930466.1 task 0: exited abnormally
srun: Terminating job step 3930466.1
```

or

```
bash-4.2$ flux mini submit -N 1 -n 1 ./datastore.py 
185925107712
bash-4.2$ flux-broker: tagpool.c:143: tagpool_alloc: Assertion `t->magic == 0x34447ff2' failed.
srun: error: quartz16: task 0: Aborted (core dumped)
srun: First task exited 30s ago
srun: step:3930492.1 tasks 1-2: running
srun: step:3930492.1 task 0: exited abnormally
srun: Terminating job step 3930492.1
srun: Job step aborted: Waiting up to 62 seconds for job step to finish.
srun: error: quartz18: task 2: Killed
srun: error: quartz17: task 1: Killed
```

**Comment by grondo** [comment]
>  flux-broker: tagpool.c:143: tagpool_alloc: Assertion `t->magic == 0x34447ff2' failed.
> srun: error: quartz16: task 0: Aborted (core dumped)

This looks like memory corruption. Can you post the stack trace, and perhaps rerun the test under valgrind?

**Comment by grondo** [comment]
> @grondo when running flux job attach -XE <jobid>, datastore.py gets here:

Thanks. Looks like the job has been started by flux, but it isn't doing anything. Adding some printf debugging to `datastore.py` might help.

**Comment by dongahn** [comment]
> This looks like memory corruption. Can you post the stack trace, and perhaps rerun the test under valgrind?

@cmoussa1: `%ulimit -c unlimited` before running this should allow you to generate a untruncated corefile. On the corefile, `%file corefile` will get you the executable path.

% gdb <broker_path> corefile

Then, at the gdb prompt: either type in `bt` or `where` will print out the backtrace.

**Comment by dongahn** [comment]
If you want to run the flux start under the control of valgrind's memcheck, one way to do this is:

```
%srun ... flux start --warp=/usr/tce/bin/memcheck_all ...
```

Then, as you run this example, valgrind will dump its data to a file (.mc) file.  

At the end of the run (either crash or manual exit), you can pretty much follow the printed instructions to view the valgrind data with memcheckview.

**Comment by cmoussa1** [comment]
Thanks to the debugging help from @dongahn, we were able to narrow this issue down:

In `*getctx (flux_t *h)` I was initially assigning `ctx->h` to `h` before allocating its memory:

```
struct conduit_ctx *ctx = flux_aug_get (h, "conduit");
ctx-> h = h;
if (!ctx) {
    ctx = calloc (1, sizeof (*ctx));
    ctx-> connected = false;
    ctx-> handlers = NULL;
    .
    .
    .
```

moving the assignment inside the `if` condition fixed the segfault issue:

```
struct conduit_ctx *ctx = flux_aux_get (h, "conduit");
if (!ctx) {
    char *user = getenv ("USER");
    ctx = calloc (1, sizeof (*ctx));
    ctx->h = h;
    ctx->connected = false;
    ctx->handlers = NULL;
    .
    .
    .
```

**Comment by cmoussa1** [comment]
This issue has been resolved and changes have been merged (thanks to PR #44), so I'll go ahead and close this.
